src/Time/TimeSequence.cpp

```diff
@@ -0,0 +1,102 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Time/TimeSequence.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <pup_stl.h>
+#include <utility>
+
+#include "ErrorHandling/Assert.hpp"
+
+namespace TimeSequences {
+template <typename T>
+EvenlySpaced<T>::EvenlySpaced(const T interval, const T offset,
+                              const OptionContext& context)
+    : interval_(static_cast<SignedT>(interval)),
+      offset_(static_cast<SignedT>(offset)) {
+  if (interval_ == 0) {
+    PARSE_ERROR(context, "Interval must be positive.");
+  }
+}
+
+template <typename T>
+std::array<std::optional<T>, 3> EvenlySpaced<T>::times_near(
+    const T time) const noexcept {
+  auto sequence_number =
+      std::floor((time + 0.5 * interval_ - offset_) / interval_);
+
+  if (not std::is_signed_v<T> and offset_ + interval_ * sequence_number < 0) {
+    ++sequence_number;
+  }
+
+  const auto previous = offset_ + interval_ * (sequence_number - 1);
+  const auto current = offset_ + interval_ * sequence_number;
+  const auto next = offset_ + interval_ * (sequence_number + 1);
+
+  ASSERT(std::is_signed_v<T> or current >= 0,
+         "Generated a negative value for an unsigned sequence.");
+
+  return {{std::is_signed_v<T> or previous >= 0 ? previous : std::optional<T>{},
+           current, next}};
+}
+
+template <typename T>
+void EvenlySpaced<T>::pup(PUP::er& p) noexcept {
+  p | interval_;
```

Comment:
If I remember correctly you should call `TimeSequence<T>::pup(p);` first. https://charm.readthedocs.io/en/latest/charm++/manual.html#allocation-outside-pup